health-monitors/syslog-health-monitor/pkg/gpufallen/gpufallen_handler.go

```diff
@@ -0,0 +1,245 @@
+// This prevents unbounded memory growth from XIDs on GPUs that never experience fallen-off errors.
+// The goroutine stops when the context is cancelled (via Close method).
+func (h *GPUFallenHandler) cleanupExpiredXIDs(ctx context.Context) {
+	// Calculate cleanup interval proportional to xidWindow (xidWindow / 5)
+	// This ensures tests with short windows don't wait too long for cleanup
+	h.mu.RLock()
+	cleanupInterval := h.xidWindow / 5
+	h.mu.RUnlock()
+
+	ticker := time.NewTicker(cleanupInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			// Context cancelled, stop cleanup goroutine
+			return
+		case <-ticker.C:
+			h.mu.Lock()
+			now := time.Now()
+			window := h.xidWindow // Read current window value
+
+			for pciAddr, record := range h.recentXIDs {
+				if now.Sub(record.timestamp) >= window {
+					delete(h.recentXIDs, pciAddr)
+				}
+			}
+			h.mu.Unlock()
+		}
+	}
+}
+
+func (h *GPUFallenHandler) parseGPUFallenError(message string) *gpuFallenErrorEvent {
+	// First check if this message itself contains "Xid" in same message
+	// If it has XID error it should be handled by XID handler
+	if common.XIDPattern.MatchString(message) {
+		return nil
+	}
+
+	m := reGPUFallenPattern.FindStringSubmatch(message)
+	if len(m) < 2 {
+		return nil
+	}
+
+	pciAddr := m[1]
+
+	// Check if this PCI address has had a recent XID error
+	// If so, skip generating an event to avoid duplicates with XID handler
+	if h.hasRecentXID(pciAddr) {
+		return nil
+	}
+
+	// Try to extract PCI ID if present in the message
+	pciID := ""
+	pciIDMatch := rePCIIDPattern.FindStringSubmatch(message)
+
+	if len(pciIDMatch) >= 2 {
+		pciID = pciIDMatch[1]
+	}
+
+	return &gpuFallenErrorEvent{
+		pciAddr: pciAddr,
+		pciID:   pciID,
+		message: message,
+	}
+}
+
+func (h *GPUFallenHandler) createHealthEventFromError(event *gpuFallenErrorEvent) *pb.HealthEvents {
+	entitiesImpacted := []*pb.Entity{
+		{EntityType: "PCI", EntityValue: event.pciAddr},
+	}
+
+	// If PCI ID is there, add it as well
+	if event.pciID != "" {
+		entitiesImpacted = append(entitiesImpacted, &pb.Entity{
+			EntityType: "PCI_ID", EntityValue: event.pciID,
+		})
+	}
+
+	// Increment metrics (node-level only to avoid cardinality explosion)
+	gpuFallenCounterMetric.WithLabelValues(h.nodeName).Inc()
+
+	healthEvent := &pb.HealthEvent{
+		Version:            1,
+		Agent:              h.defaultAgentName,
+		CheckName:          h.checkName,
+		ComponentClass:     h.defaultComponentClass,
+		GeneratedTimestamp: timestamppb.New(time.Now()),
+		EntitiesImpacted:   entitiesImpacted,
+		Message:            event.message,
+		IsFatal:            true, // GPU falling off the bus is always fatal
+		IsHealthy:          false,
+		NodeName:           h.nodeName,
+		RecommendedAction:  pb.RecommendedAction_RESTART_BM,
+		ErrorCode:          []string{"GPU_FALLEN_OFF_BUS"},
+		Metadata: map[string]string{
+			"JOURNAL_MESSAGE": event.message,
```

Comment:
nit: since we already have the message and pci address in the message already, do we want it in the metadata too?